Docs dcp (#8276)
solo-io/gloo: docs/content/guides/traffic_management/listener_configuration/hybrid_gateway.md

This is expected since the IP of our client is not `1.2.3.4`.

## Pass TLS traffic through for deprecated cipher suites {#tls-passthrough}

{{% notice note %}}

> This feature requires a Gloo Edge Enterprise license. 

Reviewer: Is there a particular version required, such as 1.14 or later?

         labels:
           run: my-nginx
           sidecar.istio.io/inject: "true"
       spec:
         containers:
         - name: my-nginx
           image: nginx
           ports:
           - containerPort: 443
           volumeMounts:
           - name: nginx-config
             mountPath: /etc/nginx
             readOnly: true
           - name: nginx-server-certs
             mountPath: /etc/nginx-server-certs
             readOnly: true
         volumes:
         - name: nginx-config
           configMap:
             name: nginx-configmap
         - name: nginx-server-certs
           secret:
             secretName: nginx-server-certs
   EOF
   ```

7. Verify that an upstream was created for your nginx server.
   ```sh
   kubectl get upstream default-my-nginx-443 -n gloo-system 
   ```

8. Create a gateway resource and specify the deprecated cipher suites for which you want to pass TLS traffic through to your upstream in the `passthroughCipherSuites` field of your `tcpGateway`. You can optionally log the ciphers that were sent by the clients by adding the `spec.options.accessLoggingService` section to your gateway configuration as shown in the following example. 

>    ```yaml

Reviewer: possibly use highlight? 
```suggestion
   {{< highlight yaml "hl_lines=12-16 31-35" >}}
```

   kubectl apply -f- <<EOF
   apiVersion: gateway.solo.io/v1
   kind: Gateway
   metadata:
     name: gateway-proxy-ssl
     namespace: gloo-system
   spec:
     bindAddress: '::'
     bindPort: 8443
     options:
       accessLoggingService:
         accessLog:
         - fileSink:
             path: /dev/shm/access_log.txt
             stringFormat: >
               [%START_TIME%] %REQUESTED_SERVER_NAME% %FILTER_STATE(io.solo.cipher_detection_info)%

>      hybridGateway:

Reviewer: Just wanted to confirm this `hybridGateway` config is correct, since we said earlier that hybrid gateway would be in a separate section?

       matchedGateways:
       - tcpGateway:
           options:
             tcpProxySettings:
               accessLogFlushInterval: 2s
           tcpHosts:
           - destination:
               single:
                 upstream:
                   name: default-my-nginx-443
                   namespace: gloo-system
             name: one
         matcher:
           passthroughCipherSuites:
           - ECDHE-RSA-AES256-SHA384
           - ECDHE-RSA-AES128-SHA256
           - AES256-SHA256
           - AES128-SHA256
           sslConfig:
             sniDomains:
             - nginx.example.com
     proxyNames:
     - gateway-proxy
     useProxyProto: false
   EOF

>    ```

Reviewer: ```suggestion
   {{< /highlight >}}
```

9. Get the IP address of the gateway.
   ```sh
   export GATEWAY_IP=$(kubectl get svc -n gloo-system gateway-proxy -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
   ```

> 11. Send a request to the gateway. In your output, verify that you see the TLS handshake between the client and the gateway and that the cipher suite that you sent as part of the request is accepted. In addition, verify that you get back a 200 HTTP response code from the NGINX server and that you see the NGINX server certificate in your CLI output. 

Reviewer: nit, the numbering skips from 9 to 11

    ```

12. Send another request to the gateway. This time, you provide a cipher that is not listed in the deprecated cipher suites. Note that the TLS connection fails as no supported cipher can be found in the request. 
    ```sh
    curl -vi --resolve nginx.example.com:443:${GATEWAY_IP} --cacert example_certs/nginx.example.com.crt --cipher "AES128" "https://nginx.example.com:443" 
    ```

    Example output: 
    ```
    * Added nginx.example.com:443:34.134.215.40 to DNS cache
    * Hostname nginx.example.com was found in DNS cache
    *   Trying 34.134.215.40:443...
    * Connected to nginx.example.com (34.134.215.40) port 443 (#0)
    * ALPN, offering h2
    * ALPN, offering http/1.1
    * Cipher selection: AES128
    * successfully set certificate verify locations:
    *  CAfile: example_certs/nginx.example.com.crt
    *  CApath: none
    * TLSv1.2 (OUT), TLS handshake, Client hello (1):
    * LibreSSL SSL_connect: Connection reset by peer in connection to nginx.example.com:443 
    * Closing connection 0
    curl: (35) LibreSSL SSL_connect: Connection reset by peer in connection to nginx.example.com:443 
    ```

## Hybrid Gateway Delegation

{{% notice note %}}

> This feature is available in Gloo Edge version 1.10.x and later.

Reviewer: Not sure we need this note anymore, because we only support versions after 1.10 anyways.

This is expected since the IP of our client is not `1.2.3.4`.

## Pass TLS traffic through for deprecated cipher suites {#tls-passthrough}

> By default, Gloo Edge supports cipher suites that are available in Envoy. If you have an upstream service that can accept only cipher suites that are deprecated or not available in Envoy, you can configure Gloo Edge to pass TLS traffic through to your upstream directly. Because TLS traffic is not terminated at the gateway, the upstream service must be capable of terminating and unencrypting the incoming TLS connection.

Reviewer 2: ```suggestion
By default, Gloo Edge only supports cipher suites that are available in Envoy, which uses [BoringSSL](https://github.com/google/boringssl) for TLS. If you provide services to clients that do not yet have support for ciphers that have been deprecated by BoringSSL, you can configure Gloo Edge to pass TLS traffic through to your upstream directly. Because TLS traffic is not terminated at the gateway, the upstream service must be capable of terminating and unencrypting the incoming TLS connection.
```

   ```shell
   mkdir example_certs
   openssl req -x509 -sha256 -nodes -days 365 -newkey rsa:2048 -subj '/O=example Inc./CN=example.com' -keyout example_certs/example.com.key -out example_certs/example.com.crt
   ```
   
2. Create a server certificate and private key for the `nginx.example.com` domain. 
   ```
   openssl req -out example_certs/nginx.example.com.csr -newkey rsa:2048 -nodes -keyout example_certs/nginx.example.com.key -subj "/CN=nginx.example.com/O=some organization"
   openssl x509 -req -sha256 -days 365 -CA example_certs/example.com.crt -CAkey example_certs/example.com.key -set_serial 0 -in example_certs/nginx.example.com.csr -out example_certs/nginx.example.com.crt
   ```

3. Create a secret that stores the certificate and key for the NGINX server. 
   ```shell
   kubectl create secret tls nginx-server-certs \
    --key example_certs/nginx.example.com.key \
    --cert example_certs/nginx.example.com.crt
   ```
   
4. Prepare your NGINX configuration. The following example configures NGINX for HTTPS traffic with the certificate that you created earlier.

>    ```shell

Reviewer 2: ```suggestion
   ```yaml
```

   cat <<\EOF > ./nginx.conf
   events {
   }

   http {
     log_format main '$remote_addr - $remote_user [$time_local]  $status '
     '"$request" $body_bytes_sent "$http_referer" '
     '"$http_user_agent" "$http_x_forwarded_for"';
     access_log /var/log/nginx/access.log main;
     error_log  /var/log/nginx/error.log;

     server {
       listen 443 ssl;

       root /usr/share/nginx/html;
       index index.html;

       server_name nginx.example.com;
       ssl_certificate /etc/nginx-server-certs/tls.crt;
       ssl_certificate_key /etc/nginx-server-certs/tls.key;
     }
   }
   EOF
   ```
   
5. Store the NGINX configuration in a configmap. 
   ```shell
   kubectl create configmap nginx-configmap --from-file=nginx.conf=./nginx.conf
   ```

6. Deploy the NGINX server. 

>    ```shell

Reviewer 2: ```suggestion
   ```yaml
```

           sidecar.istio.io/inject: "true"
       spec:
         containers:
         - name: my-nginx
           image: nginx
           ports:
           - containerPort: 443
           volumeMounts:
           - name: nginx-config
             mountPath: /etc/nginx
             readOnly: true
           - name: nginx-server-certs
             mountPath: /etc/nginx-server-certs
             readOnly: true
         volumes:
         - name: nginx-config
           configMap:
             name: nginx-configmap
         - name: nginx-server-certs
           secret:
             secretName: nginx-server-certs
   EOF
   ```

7. Verify that an upstream was created for your nginx server.
   ```sh
   kubectl get upstream default-my-nginx-443 -n gloo-system 
   ```

8. Create a gateway resource and specify the deprecated cipher suites for which you want to pass TLS traffic through to your upstream in the `passthroughCipherSuites` field of your `tcpGateway`. 
   ```yaml
   kubectl apply -f- <<EOF
   apiVersion: gateway.solo.io/v1
   kind: Gateway
   metadata:
     name: gateway-proxy-ssl
     namespace: gloo-system
   spec:
     bindAddress: '::'
     bindPort: 8443
     options:

>        accessLoggingService:

Reviewer 2: May want to add a comment that the `accessLoggingService` object here is optional for logging of the ciphers that were sent by the clients
